[mkdocs] docs: add cross chain swap tutorial and concept (#1925)
symbol/symbol: mkdocs/pages/en/devbook/transactions/cross-chain-swap.md

---
title: Cross-Chain Swap
---

# Cross-Chain Swap Between Symbol and Ethereum

> Two parties, Alice and Bob, want to exchange 0.01 ETH (on Ethereum) for 1 <XYM:> (on Symbol) without trusting each other

Reviewer: This is a specific tutorial, so you can give more details than in the textbook: What does Alice send on what chain, and what does Bob send on what chain?
In this way, it'll be easier to follow what comes next, and what accounts need to be funded with what currency.

You can probably reuse the diagram from the textbook, using specific names for the chains and the tokens being transferred.

or using an intermediary.

> Since these tokens exist on two separate blockchains, a direct transfer is not possible.

Reviewer: Maybe mention that if both tokens were on Symbol only an aggregate would be needed (and link to the atomic swap tutorial).

> This tutorial shows how to perform this <cross-chain swap:> using an

Reviewer: Since there are multiple chains involved in this tutorial, it's important that the intro explains what the code is going to do.
In this way the reader will understand why they need to install the Ethers library.

I suppose it's enough to list the 4 steps from the textbox (Alice locks, Bob locks, Alice withdraws, Bob withdraws).
Since the process is long, it would benefit from numbering these steps and use these numbers in the code comments and in the Code Explanation sections.

* Create two Symbol <accounts:>, one for Alice and one for Bob.
    See [Creating an Account from a Private Key](../accounts/create-from-private-key.md).
* Obtain XYM for Bob's account to pay for the secret lock transaction fee and locked amount.
    See [Getting Testnet Funds from the Faucet](../accounts/testnet-faucet.md).
* Create two Ethereum accounts, one for Alice and one for Bob.
    You can use [Foundry](https://book.getfoundry.sh/getting-started/installation)'s `cast wallet new` command or any
    Ethereum wallet such as MetaMask.
* Have Sepolia testnet ETH in both Ethereum accounts to pay for gas fees and enough in Alice's account to fund the HTLC.
    Sepolia ETH can be obtained from the
    [Google Cloud faucet](https://cloud.google.com/application/web3/faucet/ethereum/sepolia) or any other Ethereum
    testnet faucet.

* Install the Ethereum library for your language:

    === ":simple-python: Python"

        ```bash
        pip install web3
        ```

    === ":simple-javascript: JavaScript"

        ```bash
        npm install ethers
        ```

For background on the HTLC protocol, timing constraints, and limitations, see the

> [Cross-Chain Swaps](../../textbook/cross-chain-swaps.md) concept page.

Reviewer: This paragraph belongs to the intro.

## Full Code

{% import 'tutorial.jinja2' as tutorial with context %}

{{ tutorial.code_full('devbook/transactions/cross-chain-swap', ['py', 'js']) }}

In practice, both parties (Alice and Bob) would each run their own part on different machines.

> This tutorial combines both sides in a single script for simplicity.

Reviewer: This paragraph also belongs to the intro, right after you explain what the code does.

Author: I moved some paragraphs to the intro, while others I kept them under "Code explanation".

The intro is now concise, focusing on the overall objective. I moved the detailed explanation to the "Code explanation" section to make the tutorial easier to approach and to introduce more complex concepts (e.g. tx flow) as they are used, once they've seen the code.

The code defines helper functions to fetch the network time and fees, announce transactions, and poll for confirmation,
following the same patterns described in the [Transfer](./transfer.md) tutorial.

> This tutorial does not wait for transaction [finality](../../textbook/cross-chain-swaps.md#finality) between steps,

Reviewer: Better to use `<finalization:|finality>` link.

which a production implementation must do to prevent rollback-related risks.

## Ethereum HTLC Contract

This tutorial uses a sample HTLC contract deployed on Ethereum as the counterpart to Symbol's secret lock.
The contract source is available in the

> [hashed-timelock-contract-ethereum](https://github.com/gimre-xymcity/hashed-timelock-contract-ethereum)

Author: @segfaultxavi  The ethereum contract is under gimre's personal account. We might want to move it under symbol/symbol or create a new repo. thoughts?

Reviewer: I'm OK with leaving it there, but happy to hear from @gimre-xymcity, @Jaguar0625, or @0x6861746366574

Reviewer: We'll move it to the Symbol organization. When that happens we'll update the link here.

Reviewer: Actually, that's already been done and I've updated the link.

repository.

!!! warning "Educational use only"
    Any contract used in production must carefully calibrate lock and contract expiry times, as timing is critical for
    the security of both parties.

The contract provides three key methods:

* `newContract(address receiver, bytes32 hashlock, uint timelock)`: Creates a new HTLC with a recipient, hashlock, and
    a Unix timestamp as timelock.
    Comparable to Symbol's <ser:SecretLockTransactionV1>.
* `withdraw(bytes32 contractId, bytes proof)`: Allows the recipient to claim funds by providing the proof that matches
    the hashlock.
    Comparable to Symbol's <ser:SecretProofTransactionV1>.
* `refund(bytes32 contractId)`: Returns funds to the creator after the timelock expires.
    In Symbol, refunds happen automatically when a secret lock expires.

The contract has been deployed on the Sepolia testnet at address `0xd58e030bd21c7788897aE5Ea845DaBA936e91D2B`.

## Code Explanation

### Setting Up Accounts

{{ tutorial.code_snippet(['py:154:182', 'js:124:156']) }}

Both parties need accounts on both chains.
Alice uses her Ethereum account to lock ETH and her Symbol account to claim XYM.

> Bob uses his Symbol account to lock XYM and his Ethereum account to withdraw ETH.

Reviewer: This must be explained in the intro (as previously commented).
Also, I still don't know who initiates the transfer.

    the swap.
* **Value:** 0.01 ETH sent along with the transaction.

The transaction receipt contains a `LogHTLCNew` event with a `contractId` that identifies this HTLC.
Bob will need this `contractId` later to withdraw the ETH.

### Bob: Creating a Secret Lock on Symbol

{{ tutorial.code_snippet(['py:222:246', 'js:194:222']) }}

Bob first queries the Ethereum HTLC contract using `getContract` to retrieve the hashlock that Alice used.

!!! note "Verify before locking"
    Bob should verify the full contract details (amount, recipient, timelock) before locking his own funds.
    This tutorial only reads the hashlock for simplicity.

Bob then creates a <ser:SecretLockTransactionV1> on Symbol, locking 1 XYM for Alice, using the **same hashlock**:

* **Recipient:** Alice's Symbol address.
* **Mosaic:** 1 XYM (expressed as `1_000000` atomic units with divisibility 6).
* **Duration:** 5760 blocks (~48 hours at 30-second block times).

    !!! warning "Timelock ordering"
        This duration must be **shorter** than Alice's 72-hour Ethereum timelock.
        Otherwise, Alice could refund her ETH and still claim Bob's XYM.

>         The gap between the two must be **large** enough: is the safety margin that guarantees Bob can still withdraw

Reviewer: ```suggestion
        The gap between the two must be **large** enough: it is the safety margin that allows Bob to withdraw
```

        on Ethereum even if Alice reveals the proof at the last moment.
        See [Timing Constraints](../../textbook/cross-chain-swaps.md).

* **Hashlock (`secret` field):** The hashlock retrieved from the Ethereum contract.
* **Hash algorithm:** `hash_256` (double SHA-256), must match the algorithm used in the counterpart chain's HTLC.

### Alice: Claiming XYM on Symbol

{{ tutorial.code_snippet(['py:263:275', 'js:239:254']) }}

> Once Bob's secret lock is confirmed, Alice can claim the locked XYM by revealing the proof.

Reviewer: You should say that the full code signs, announces, and confirms Bob's lock transaction, but that it has been omitted from the snippet above.

Reviewer: Also, I would insist that "Alice can claim the locked XYM **on Symbol**" (as you did in the header).
There's too many chains and too many accounts involved. It doesn't hurt to be extra clear :D

She creates a <ser:SecretProofTransactionV1> with:

* **Recipient:** Alice's own Symbol address (the same address set in Bob's secret lock).
* **Hashlock (`secret` field):** The same hashlock used in the secret lock.
* **Hash algorithm:** `hash_256` (must match the secret lock).
* **Proof:** The original random bytes that Alice generated.

> Once this transaction is announced and confirmed, the proof is **publicly visible** on the Symbol blockchain.

Reviewer: And, more importantly, Alice gets her XYM.

Bob (or anyone) can read it from the transaction data.

### Bob: Withdrawing ETH on Ethereum

{{ tutorial.code_snippet(['py:292:316', 'js:272:288']) }}

> After Alice's secret proof is confirmed, Bob retrieves the proof from the confirmed transaction using the

Reviewer: How does Bob know `proof_hash`? Does Alice need to send it to him, or can he find out without her help? Does he need to monitor the chain looking for a `SecretProofTransactionV1` with the correct `Hashlock`?

Author: Edited the code. Now we are retrieving it from the chain by monitoring Alice's confirmed txs.

<get:/transactions/confirmed/{transactionId}> endpoint and calls `withdraw` on the Ethereum HTLC contract with two
arguments:

* **Contract ID:** The HTLC identifier from the `LogHTLCNew` event emitted when Alice locked the ETH.
* **Proof:** The proof Alice revealed on Symbol.

!!! warning "Withdrawal deadline"

>     Bob must complete this before Alice's Ethereum timelock expires.
>     After that, Alice can call `refund` on the Ethereum contract and reclaim her ETH.

Reviewer: ```suggestion
    Bob must complete this step before Alice's Ethereum timelock expires.
    Once expired, Alice can call `refund` on the Ethereum contract and reclaim her ETH.
```

"Naked this" is a problem in English, and it was not clear if Alice can refund after the timelock expires or after Bob completes this step.

> Once the Ethereum transaction is confirmed, the swap is complete: Alice has 1 XYM and Bob has 0.01 ETH.

Reviewer: Again, let's be extra clear here: once this eth transaction is complete, Bob will receive Alice's 0.01 ETH.
Alice already received Bob's 1 XYM at the end of step 3.

## Output

The output shown below corresponds to a typical run of the program.

```text linenums="1" hl_lines="9 10 16 19 50 78 81 82"
--8<-- 'devbook/transactions/cross-chain-swap.log'
```

Key points in the output:

* **Lines 9-10:** Alice generates the proof and hashlock. The proof must remain secret until Alice reveals it.
* **Line 16:** The HTLC contract ID identifies Alice's Ethereum lock. Bob uses this to query the hashlock and later
    to withdraw.
* **Line 19:** Bob retrieves the hashlock from the Ethereum contract using `getContract`.
* **Line 50:** Bob's Symbol secret lock is confirmed. Alice can now claim the XYM.

> * **Line 78:** Alice's secret proof is confirmed, revealing the proof on the Symbol blockchain.

Reviewer: Again, we have to say that Alice has just received the XYM.
